File: eventstore/namespace/namespace.go
// Package namespace provides per-project segregation for any
// eventstore.EventStore (architecture R9). It is a single decorator over the
// EventStore port: it prefixes every stream a project writes with the project's
// namespace and scopes every global read and subscription to that namespace, so
// one backend (a shared SQLite file or a shared KurrentDB instance) can hold many
// projects without their streams or graphs ever mixing.
//
// Because it depends only on the port, it is written once and wraps every
// backend; that is dependency inversion buying the single implementation.
package namespace

import (
	"context"
	"strings"
	"sync"

	"github.com/virtual-velocitation/rigger/eventstore"
)

// Store wraps an EventStore so all of its data is scoped to one project. Callers
// use plain, unprefixed stream names and never see the namespace.
type Store struct {
	inner  eventstore.EventStore
	prefix string
}

var _ eventstore.EventStore = (*Store)(nil)

// New scopes inner to the named project.
func New(inner eventstore.EventStore, project string) *Store {
	return &Store{inner: inner, prefix: "proj-" + project + "-"}
}

// Append writes to the project-scoped stream.
func (s *Store) Append(ctx context.Context, stream string, expected eventstore.ExpectedRevision, events ...eventstore.Event) (eventstore.Position, error) {
	return s.inner.Append(ctx, s.prefix+stream, expected, events...)
}

// ReadStream reads one project-scoped stream, returning clean stream names.
func (s *Store) ReadStream(ctx context.Context, stream string, from eventstore.Revision, dir eventstore.Direction) ([]eventstore.Event, error) {
	evs, err := s.inner.ReadStream(ctx, s.prefix+stream, from, dir)
	return s.strip(evs), err
}

// ReadAll reads only this project's events, in global order.
func (s *Store) ReadAll(ctx context.Context, from eventstore.Position, dir eventstore.Direction, filter eventstore.Filter) ([]eventstore.Event, error) {
	evs, err := s.inner.ReadAll(ctx, from, dir, s.scopeFilter(filter))
	return s.strip(evs), err
}

// SubscribeAll subscribes to only this project's events.
func (s *Store) SubscribeAll(ctx context.Context, from eventstore.Position, filter eventstore.Filter) (eventstore.Subscription, error) {
	inner, err := s.inner.SubscribeAll(ctx, from, s.scopeFilter(filter))
	if err != nil {
		return nil, err
	}
	return newStrippingSub(inner, s.prefix), nil
}

// Close closes the wrapped store.
func (s *Store) Close() error { return s.inner.Close() }

// scopeFilter forces the project namespace, composing it with any caller prefix
// (interpreted within the namespace).
func (s *Store) scopeFilter(f eventstore.Filter) eventstore.Filter {
	return eventstore.Filter{StreamPrefix: s.prefix + f.StreamPrefix}
}

func (s *Store) strip(evs []eventstore.Event) []eventstore.Event {
	for i := range evs {
		evs[i].Stream = strings.TrimPrefix(evs[i].Stream, s.prefix)
	}
	return evs
}

// strippingSub wraps a subscription to strip the namespace prefix from each
// delivered event.
type strippingSub struct {
	inner     eventstore.Subscription
	prefix    string
	out       chan eventstore.Event
	done      chan struct{}
	closeOnce sync.Once
}

var _ eventstore.Subscription = (*strippingSub)(nil)

func newStrippingSub(inner eventstore.Subscription, prefix string) *strippingSub {
	sub := &strippingSub{
		inner:  inner,
		prefix: prefix,
		out:    make(chan eventstore.Event, 128),
		done:   make(chan struct{}),
	}
	go sub.run()
	return sub
}

func (sub *strippingSub) run() {
	defer close(sub.out)
	in := sub.inner.Events()
	for {
		select {
		case e, ok := <-in:
			if !ok {
				return
			}
			e.Stream = strings.TrimPrefix(e.Stream, sub.prefix)
			select {
			case sub.out <- e:
			case <-sub.done:
				return
			}
		case <-sub.done:
			return
		}
	}
}

func (sub *strippingSub) Events() <-chan eventstore.Event { return sub.out }
func (sub *strippingSub) Err() error                      { return sub.inner.Err() }

func (sub *strippingSub) Close() error {
	sub.closeOnce.Do(func() { close(sub.done) })
	return sub.inner.Close()
}

File: src/eventstore/namespace.rs
//! Per-project segregation for any EventStore: a single decorator over the port
//! that prefixes every stream a project writes and scopes every global read and
//! subscription to that namespace, so one backend (a shared SQLite file or a
//! shared KurrentDB instance) can hold many projects without their streams ever
//! mixing. Callers use plain, unprefixed stream names and never see the namespace.
//!
//! Because it depends only on the port, it is written once and wraps every
//! backend - dependency inversion buying the single implementation.

use super::{
    Direction, Error, Event, EventStore, ExpectedRevision, Filter, Position, Subscription,
};

/// Namespaced wraps an EventStore so all of its data is scoped to one project.
pub struct Namespaced<'a> {
    inner: &'a dyn EventStore,
    prefix: String,
}

impl<'a> Namespaced<'a> {
    /// Scope inner to the named project.
    pub fn new(inner: &'a dyn EventStore, project: &str) -> Self {
        Namespaced {
            inner,
            prefix: format!("proj-{project}-"),
        }
    }

    fn scoped(&self, stream: &str) -> String {
        format!("{}{stream}", self.prefix)
    }

    /// Force the project namespace, composing it with any caller prefix
    /// (interpreted within the namespace).
    fn scope_filter(&self, filter: &Filter) -> Filter {
        let caller = filter.stream_prefix.as_deref().unwrap_or("");
        Filter {
            stream_prefix: Some(format!("{}{caller}", self.prefix)),
        }
    }
}

impl EventStore for Namespaced<'_> {
    fn append(
        &self,
        stream: &str,
        expected: ExpectedRevision,
        events: &[Event],
    ) -> Result<Position, Error> {
        self.inner.append(&self.scoped(stream), expected, events)
    }

    fn read_stream(
        &self,
        stream: &str,
        from: Position,
        dir: Direction,
    ) -> Result<Vec<Event>, Error> {
        self.inner.read_stream(&self.scoped(stream), from, dir)
    }

    fn read_all(
        &self,
        from: Position,
        dir: Direction,
        filter: &Filter,
    ) -> Result<Vec<Event>, Error> {
        self.inner.read_all(from, dir, &self.scope_filter(filter))
    }

    fn subscribe_all(&self, from: Position, filter: &Filter) -> Result<Subscription, Error> {
        self.inner.subscribe_all(from, &self.scope_filter(filter))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::eventstore::sqlite::Store;

    #[test]
    fn segregates_projects_in_one_backend() {
        let backend = Store::open(":memory:").unwrap();
        let alpha = Namespaced::new(&backend, "alpha");
        let beta = Namespaced::new(&backend, "beta");

        // Both projects write to a stream named "run".
        alpha
            .append(
                "run",
                ExpectedRevision::Any,
                &[Event::new("A1", b"a".to_vec())],
            )
            .unwrap();
        beta.append(
            "run",
            ExpectedRevision::Any,
            &[Event::new("B1", b"b".to_vec())],
        )
        .unwrap();

        // Each project's global read sees only its own events.
        let a = alpha
            .read_all(0, Direction::Forward, &Filter::default())
            .unwrap();
        assert_eq!(
            a.iter().map(|e| e.type_.as_str()).collect::<Vec<_>>(),
            ["A1"]
        );
        let b = beta
            .read_all(0, Direction::Forward, &Filter::default())
            .unwrap();
        assert_eq!(
            b.iter().map(|e| e.type_.as_str()).collect::<Vec<_>>(),
            ["B1"]
        );

        // The same stream name in different projects does not collide.
        let a_run = alpha.read_stream("run", 0, Direction::Forward).unwrap();
        assert_eq!(a_run.len(), 1);
        assert_eq!(a_run[0].type_, "A1");
    }

    #[test]
    fn passes_the_contract() {
        let backend = Store::open(":memory:").unwrap();
        let scoped = Namespaced::new(&backend, "contract");
        crate::eventstore::contract::assert_contract(&scoped);
    }
}
